src/app/labeling/qgsmaptoolmovelabel.cpp

```diff
@@ -295,124 +381,190 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
+
+        if ( !isCalloutMove && isCurvedOrLine && !currentLabelDataDefinedLineAnchorPercent( lineAnchorPercentOrig, lineAnchorPercentSuccess, lineAnchorPercentCol,
+             lineAnchorClippingOrig, lineAnchorClippingSuccess, lineAnchorClippingCol,
+             lineAnchorTypeOrig, lineAnchorTypeSuccess, lineAnchorTypeCol,
+             lineAnchorTextPointOrig,  lineAnchorTextPointSuccess, lineAnchorTextPointCol ) )
         {
           return;
         }
-        else if ( isCalloutMove && !currentCalloutDataDefinedPosition( xPosOrig, xSuccess, yPosOrig, ySuccess, xCol, yCol ) )
+        else if ( !isCalloutMove && !currentLabelDataDefinedPosition( xPosOrig, xSuccess, yPosOrig, ySuccess, xCol, yCol, pointCol ) )
         {
           return;
         }
-
-        double xPosNew = 0;
-        double yPosNew = 0;
-
-        if ( !xSuccess || !ySuccess )
-        {
-          xPosNew = releaseCoords.x() - mClickOffsetX;
-          yPosNew = releaseCoords.y() - mClickOffsetY;
-        }
-        else
+        else if ( isCalloutMove && !currentCalloutDataDefinedPosition( xPosOrig, xSuccess, yPosOrig, ySuccess, xCol, yCol ) )
         {
-          //transform to map crs first, because xdiff,ydiff are in map coordinates
-          const QgsMapSettings &ms = mCanvas->mapSettings();
-          const QgsPointXY transformedPoint = ms.layerToMapCoordinates( vlayer, QgsPointXY( xPosOrig, yPosOrig ) );
-          xPosOrig = transformedPoint.x();
-          yPosOrig = transformedPoint.y();
-          xPosNew = xPosOrig + xdiff;
-          yPosNew = yPosOrig + ydiff;
+          return;
         }
 
-        //transform back to layer crs
-        if ( mCanvas )
+        // Handle curved offset
+        if ( isCurvedOrLine )
         {
-          const QgsMapSettings &s = mCanvas->mapSettings();
-          const QgsPointXY transformedPoint = s.mapToLayerCoordinates( vlayer, QgsPointXY( xPosNew, yPosNew ) );
-          xPosNew = transformedPoint.x();
-          yPosNew = transformedPoint.y();
-        }
 
-        if ( !isCalloutMove )
-          vlayer->beginEditCommand( tr( "Moved label" ) + QStringLiteral( " '%1'" ).arg( currentLabelText( 24 ) ) );
-        else
-          vlayer->beginEditCommand( tr( "Moved callout" ) );
+          const QgsFeature feature { mCurrentLabel.layer->getFeature( featureId ) };
+          const QgsMapSettings &ms = mCanvas->mapSettings();
+          const QgsPointXY pointMapTransformed = ms.mapToLayerCoordinates( mCurrentLabel.layer, releaseCoords );
+          const QgsGeometry pointMapGeometry = QgsGeometry::fromPointXY( pointMapTransformed );
+          const QgsGeometry anchorPoint { feature.geometry().nearestPoint( pointMapGeometry ) };
+          const double lineAnchorPercent { feature.geometry().lineLocatePoint( anchorPoint ) / feature.geometry().length() };
```

Comment:
```suggestion
          const double lineAnchorPercent { feature.geometry().lineLocatePoint( pointMapGeometry ) / feature.geometry().length() };
```
This will automatically do the snapping, so no need for the nearestPoint call too